src/apps/alert/AlertReceiver.cc

```diff
@@ -45,17 +46,23 @@ void AlertReceiver::handleMessage(cMessage *msg)
 {
     if (msg->isSelfMessage())
         return;
-    AlertPacket* pkt = check_and_cast<AlertPacket*>(msg);
 
-    if (pkt == 0)
-        throw cRuntimeError("AlertReceiver::handleMessage - FATAL! Error when casting to AlertPacket");
+    Packet* pPacket = check_and_cast<Packet*>(msg);
+
+    if (pPacket == 0)
+    {
+        throw cRuntimeError("AlertReceiver::handleMessage - FATAL! Error when casting to inet packet");
+    }
+
+    // read Alert header
+    auto alert = pPacket->popAtFront<AlertPacket>();
 
     // emit statistics
-    simtime_t delay = simTime() - pkt->getTimestamp();
+    simtime_t delay = simTime() - alert->getTimestamp();
```

Comment:
Using getTimestamp for statistics is generally not a god idea, because Packet instances can be copied at any given time in the underlying network. I think you should attach a CreationTimeTag to the packet data instead.

Reply:
Good idea, we will add a CreationTimeTag (not only in the Alert-App but also all other apps).

Howevever, regarding the statistics calculation (i.e. the delay of a received packet within the receiver) my understanding of the API is the following (correct me if I am wrong): If the Packet instance would be copied in the underlying network, the timeStamp field inserted by the application would also be copied, and a _new_ CreationTimeTag might be added. However, since on the application layer we want to calculate the end-to-end delay (time difference between creation of the original packet and the reception at the application layer at receiver side), using the timeStamp inserted by the application would still be correct, wouldn't?

Looking a the other inet protocols, it seems that for calculating the delay based on the CreationTimeTag the protocols (except for the SCTP implementation) do not implement any statistics but rely on ageneric mechanism (DataAgeFilter)?

However, in order not to break existing scripts relying on these statistics reported by simulte,I would suggest the following:

1. We add CreationTimeTags in all simulte applications.
1. BUT we still calculate the application layer delay based on the application layer time stamp field (users could alternatively use the generic filters if they want the delay calculation to be based on the CreationTimeTag.

Would that be ok in your opinion?

Reply:
The CreationTimeTag attached to the data will be copied with the data transparently in all involved protocols throughout the underlying network. In contrast, the timestamp field of cMessage will not be maintained. In fact, there's no way of maintaining it correctly. Think about fragmentation and aggregation for a TCP connection, the receiver application can receive a packet that contains data which was created at different times in the sending application. There's simply no way of representing that with a simple timestamp field. So no, using the timestamp field of the cMessage is generally not correct. It may produce the same results in some simple networks, but may be completely wrong in others.

DataAgeFilter is the right mechanism for such cases. The MessageTsAgeFilter is not quite right, but what is more important is that the protocols _do not_ copy the timestamp when creating packets. For example, TCP may create a TCP segment packet from multiple packets, so it doesn't even make sense talking about copying the timestamp field.

So for point 2 I would suggest to use the DataAgeFilter instead.

Reply:
I think there is a misunderstanding (due to the maybe not optimal naming of header fields in simulte)  - getTimestamp() is not called on a cMessage, it is called on the packet header of the AlertPacket, which includes an own (application layer) timestamp:
```
class AlertPacket extends inet::FieldsChunk
{
    unsigned int sno;
    simtime_t timestamp;
}
```

Reply:
Oh, I see. But in this case the original AlertPacket definition _did hide_ the cMessage timestamp field though. The cMessage::getTimestamp() function is not virtual, so calling getTimestamp on a cPacket didn't return the application specific timestamp. One must have been very careful about this.

But you are right, apart from that, my reasoning above doesn't apply. If the application has a specific timestamp field then it's safe to use that.